op.c

```diff
@@ -4299,6 +4299,52 @@ Perl_invert(pTHX_ OP *o)
     return newUNOP(OP_NOT, OPf_SPECIAL, scalar(o));
 }
 
+/* Warn about possible precedence issues if op is a control flow operator that
+   does not terminate normally (return, exit, next, etc).
+*/
+static void
+S_check_terminates(pTHX_ OP *op)
+{
+    assert(op != NULL);
+
+    /* [perl #59802]: Warn about things like "return $a or $b", which
+       is parsed as "(return $a) or $b" rather than "return ($a or
+       $b)".
+    */
+    switch (op->op_type) {
+    case OP_NEXT:
+    case OP_LAST:
+    case OP_REDO:
+        /* XXX: Perhaps we should emit a stronger warning for these.
+           Even with the high-precedence operator they don't seem to do
+           anything sensible.
+
+           But until we do, fall through here.
+         */
+    case OP_EXIT:
+    case OP_RETURN:
+    case OP_DIE:
+    case OP_GOTO:
+        /* XXX: Currently we allow people to "shoot themselves in the
+           foot" by explicitly writing "(return $a) or $b".
+
+           Warn unless we are looking at the result from folding or if
+           the programmer explicitly grouped the operators like this.
+           The former can occur with e.g.
+
+                use constant FEATURE => ( $] >= ... );
+                sub { not FEATURE and return or do_stuff(); }
+         */
+        if (!op->op_folded && !(op->op_flags & OPf_PARENS))
+            Perl_ck_warner(aTHX_ packWARN(WARN_SYNTAX),
+                           "Possible precedence issue with control flow operator");
+        /* XXX: Should we optimze this to "return $a;" (i.e. remove
+           the "or $b" part)?
```

Comment:
Might be worth doing - someone who deparses their code might notice the discrepancy even if they missed the warning.

Reply:
+1

Reply:
This is part of the original code that I just moved into a separate function. I don't think this PR is the right place to make this kind of change.

Reply:
I have now taken a stab at this idea in https://github.com/mauke/perl5/commits/fold-flow-ops. Comments welcome.

I'm still not sure if it should be added to this PR.

Reply:
Oh yeah, fair enough, separate PR sounds good.